### DISCUSSION TITLE
How can I right pad an integer with zeros?

### DISCUSSION THREAD

**Discussion by derekmahar** [johnkerl/miller#1153] [Q&A] [answered]
How can I _left-align_ an integer and pad it to the _right_ with zeros?  For example, left-aligning 123 and padding it with zeros to its right would produce "123000".

Miller function [`fmtnum`](https://miller.readthedocs.io/en/latest/reference-dsl-builtin-functions/#fmtnum) can _right-align_ an integer and pad it with zeros to its _left_:

```
$ echo number=123 | mlr put '$result = fmtnum($number, "%06d");'
number=123,result=000123
```

Can `fmtnum` _left-align_ and _right-pad_ a number with zeros?  Can [`format`](https://miller.readthedocs.io/en/latest/reference-dsl-builtin-functions/#format) or some other function do this?

**Comment by derekmahar** [comment, marked as answer]
Turns out that I required a function that left-aligns and right-pads a _string_, not an integer.  I needed this specifically in order to left-align and right-pad fractional seconds of an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) timestamp string.  I couldn't use an integer to represent the decimal digits of the fractional seconds because fractions such as "085" or "001" are technically not integers and even if Miller could convert them to integers, it would strip the significant leading zeros.  Instead, I had to treat these fractional seconds as strings that preserved all digits, including leading zeros.

To achieve this goal, I wrote the following function `right_pad` that appends repeated copies of string `padding` to input string `unpadded_string` to a maximum length of `width`.

```
func right_pad(str unpadded_string, int width, str padding) {
  str padded_string = unpadded_string;

  for (; strlen(padded_string) <= width; padded_string .= padding) {};

  return truncate(padded_string, width);
}
```

I think Miller could benefit from `right_pad` or similar as a built-in string function.

**Reply by johnkerl** [reply]
@derekmahar great idea!! :) This is tracked in #1166.

**Comment by aborruso** [comment]
hi @derekmahar , probably It's not elegant and it's not what you're looking for.

Starting from this input file

```
a
1
12
253
426
2
```

and using powers of 10, and cell length, you could run something like

```
mlr --csv  put '$b=$a*pow(10,6-strlen(string($a)))' input.csv
```

to have 

| a | b |
| --- | --- |
| 1 | 100000 |
| 12 | 120000 |
| 253 | 253000 |
| 426 | 426000 |
| 2 | 200000 |

**Reply by derekmahar** [reply]
@aborruso, your solution certainly solves the problem that I stated in the title and I did think of this one, but I chose a more general solution that right-pads any input string with repetitions of another string.  Your powers-of-ten solution is probably much faster than my solution, though, so I should use this one specifically for integers.

**Reply by derekmahar** [reply]
Oh, I overlooked one important point.  Your solution can't right-pad an "integer" string that has leading zeroes such as "001" or "023" because Miller won't convert it to an integer unless you first strip the leading zeroes.  However, if you strip those leading zeroes, you would have to save them and then prepend them to the integer that results from multiplying the "stripped" integer by the necessary powers-of-ten.

**Reply by aborruso** [reply]
> Your solution can't right-pad an "integer" string that has leading zeroes such as "001" or "023"

True, but 001 is not an integer. But you are right, I understood you: your question is no longer that of the original subject, but it is a generic one, related to rigth-pad, and not strictly to "integer with zeros".

**Comment by amp42** [comment]
I needed to left-pad a string, here's a slight mod to @derekmahar's function:

```
func lpad(unpadded_string, int width, string padding) {
  if (strlen(string(unpadded_string)) >= width) {return string(unpadded_string);}

  str padded_string = string(unpadded_string);
  for (; strlen(padded_string) < width; padded_string = padding . padded_string) {};
  return truncate(padded_string, width);
}
```

Some of my unpadded strings could pass as numbers (`42`) and some are number-looking strings (`042`), thus type coercion via `string()` and untyped `unpadded_string` argument. I stuck the function in `$HOME/functions.mlr` and used it for left-padding to a width of 3 with zeros:

```mlr --load ~/functions.mlr --csv --from myfile.csv put '$foo=lpad($foo,3,"0")'```

**Reply by derekmahar** [reply]
Why leave `unpadded_string` as untyped when you could just require that the client coerce the argument assigned to this parameter as a string?  As a string parameter, `unpadded_string` could handle both number and number-looking strings that have leading zeros or the client might decide to strip leading zeros from the argument before passing it to `lpad`.

Note that unlike `right_pad`, `lpad` neglects to truncate its return value when `unpadded_string` is longer than the maximum width.

**Reply by amp42** [reply]
> Why leave `unpadded_string` as untyped when you could just require that the client coerce the argument assigned to this parameter as a string?

...because `mlr` doesn't like that with numbers. With `unpadded_string` as `str` and inlined zero for padding:

```
func lpadzero(str unpadded_string, int width) {
  if (strlen(unpadded_string) >= width) {return unpadded_string;}

  str padded_string = unpadded_string;
  for (; strlen(padded_string) < width; padded_string = 0 . padded_string) {};
  return truncate(padded_string, width);
}
```

`test.csv`:

```
foo
011
002
3
04
55
666
```

``mlr --load ~/functions.mlr --csv --from test.csv put '$foo=lpadzero($foo,3)'`` yields

`mlr: couldn't assign variable str unpadded_string from value int 3`

**Reply by derekmahar** [reply]
Right, so that's why you should instead call it using `mlr --load ~/functions.mlr --csv --from test.csv put '$foo=lpadzero(string($foo),3)'`.

**Comment by johnkerl** [comment]
@derekmahar @aborruso https://github.com/johnkerl/miller/pull/1205 is in progress
